src/algebra/module/basic.lean

```diff
@@ -509,6 +509,12 @@ section module
 
 variables [ring R] [add_comm_group M] [module R M]
 
+lemma smul_injective [no_zero_smul_divisors R M] {c : R} (hc : c ≠ 0) :
+  function.injective (λ (x : M), c • x) :=
+λ x y h, sub_eq_zero.mp ((smul_eq_zero.mp
+  (calc c • (x - y) = c • x - c • y : smul_sub c x y
+                ... = 0 : sub_eq_zero.mpr h)).resolve_left hc)
```

Comment:
Do you need `add_comm_group M`, or is `add_cancel_comm_monoid M` sufficient?

Reply:
Hmm, you mean a proof along the lines of:
```lean
λ x y h, add_left_cancel
  (calc (c - 1) • x + x = c • x : _
        ...             = c • y : h
        ...             = (c - 1) • x + y : sorry)
```
I can't figure out what to put in place of the `sorry`, though.

Reply:
Nor can I. You can relax `ring` to `semiring` without changing the proof though.